Added spotless java formatter plugin and updated readme

Shaily20 opened the pull request:
<!-- Date Here in dd mmm yyyy-->
Date: 12 Mar 2024
<!--Developer Name Here-->
Developer Name: Shaily

---

## Issue Ticket Number
<!--Issue ticket this PR closes-->
#36 

## Description

<!--Description of the changes made in this PR-->
The codebase is formatted consistently according to the defined spotless google Java Format.
The check goal fails the build in local as well in CI pipeline if the files are not formatted using mvn spotless:apply.
I have not added the goal to format the files in pom.xml, instead check goal will fail the build automatically if not formatted.
How to use the formatter is mentioned in the readme.
The *.java files have been formatted using spotless:apply command and are present in the PR as well.

### Documentation Updated?

- [x] Yes
- [ ] No

<!--Additional notes about documentation update if applicable-->

### Under Feature Flag

- [ ] Yes
- [x] No

<!--Indicate if changes are under a feature flag-->

### Database Changes

- [ ] Yes
- [x] No

<!--Notes on any database changes-->

### Breaking Changes

- [ ] Yes
- [x] No

<!--Notes on breaking changes or related issue tickets if applicable-->

### Development Tested?

- [ ] Yes
- [ ] No

<!--Confirmation of local testing during development-->

## Screenshots
<details>
<summary>Screenshot 1</summary>

<!-- Attach your screenshots here👇-->

</details>

<!--Attach or link to relevant screenshots or visual aids-->

## Test Coverage
<details>
<summary>Screenshot 1</summary>

<!-- Attach your screenshots here👇-->

</details>

<!--Attach Details on test coverage and outcomes-->

## Additional Notes

<!--Any additional notes, considerations or explanations for reviewers-->

debanjanc01 commented:
Hey @Shaily20, why not add the plugin to the compile phase? And take the goal based on environment. So if the environment is dev, the dev doesn't have to run a separate command. It will be formatted automatically.

And when the environment is not dev, the goal should be for `check`. This is what will happen for pipelines and deployment.

Shaily20 commented:
> Hey @Shaily20, why not add the plugin to the compile phase? And take the goal based on environment. So if the environment is dev, the dev doesn't have to run a separate command. It will be formatted automatically.
> 
> And when the environment is not dev, the goal should be for `check`. This is what will happen for pipelines and deployment.

Hi @debanjanc01 
I added it to the validate phase because it is always run when compile phase is executed according to the maven build lifecycle. But, I will change it to the compile phase.
Regarding the goals based on environment, I had few doubts. By environment, do you mean branches or prod/dev envs'?
Because our current CI pipeline runs when any Pull Request is created. 
The check goal is fine as it will be executed and fail the build if not formatted. But, when I add a format goal, as we discussed once that changes made in the code due to ci build will not persist. So, we will still need to run mvn compile in local before pushing changes.

satyam73 commented on .gitignore:
```diff
@@ -26,4 +26,6 @@ replay_pid*
\ No newline at end of file
+*.env
+
+target/
```
what's get created in target folder?

Shaily20 replied:
spotless-index containing information related to formatting like time of formatting

satyam73 commented on README.md:
```diff
@@ -171,6 +171,14 @@ Now you can connect to the mysql running in the docker container, also to connec
+### Code Formatting
+
+This repo uses https://github.com/diffplug/spotless/tree/main/plugin-maven#java for formatting files.
+Please run `mvn spotless:apply` before check-in to fix any formatting errors.
```
check-in means?

Shaily20 replied:
By check-in, I meant pushing the code in the repository. I'll make it more clear. Thank you

satyam73 commented on README.md:
```diff
@@ -171,6 +171,14 @@ Now you can connect to the mysql running in the docker container, also to connec
+This repo uses https://github.com/diffplug/spotless/tree/main/plugin-maven#java for formatting files.
+Please run `mvn spotless:apply` before check-in to fix any formatting errors.
+
+The Continuous Integration build for pushed commits may fail when a Pull Request is created if formatting is not applied.
```
```suggestion
The Continuous Integration build for pushed commits may fail when a Pull Request is created if your code doesn't follow project's formatting guideline.
```

satyam73 commented on skill-tree/pom.xml:
```diff
@@ -151,6 +151,43 @@
+                            <tabs>true</tabs>
+                            <spacesPerTab>2</spacesPerTab>
+                        </indent>
+                        <indent>
+                            <spaces>true</spaces>
+                            <spacesPerTab>4</spacesPerTab>
```
This part is two times here, may be because of some reason. Can you please elaborate once?

Shaily20 replied:
Hi @satyam73 
This is done to enforce 4 spaces indentation when using tabs. Since, we are using googleJavaFormat, we need to turn google's 2-space indentation to tabs, then the tabs to 4-space indentation. Hence, its written 2 times here.
For more information, the following link can be referred.
https://github.com/diffplug/spotless/issues/420#issuecomment-512350756

debanjanc01 commented:
So the goal should be dynamically changed based on the environment. Maven allows you to do that. Try exploring it.

Shaily20 commented:
> @Shaily20 So the goal should be dynamically changed based on the environment. Maven allows you to do that. Try exploring it.

Hi @debanjanc01 , I have added 2 profiles in the pom.xml where local profile is active by default & running a mvn compile will format the code when built in local. Devs do not have to call spotless:apply explicitly.
In the ci.yml file, I have added the profile using -P in the mvn verify command which will run the check goal instead. 
Please let me know if my understanding is in right direction & if any more changes are needed.
Thank you

iamitprakash commented:
I don't see any SS added for new changes

iamitprakash reviewed (changes requested):
Please resolve conflict(s)

Shaily20 commented:
Hi All,
My approvals went stale & were dismissed after resolving conflicts. Could you please provide your approvals again? 
Thank you

iamitprakash reviewed (changes requested):
quick question.. do we need to change anything into GitHub workflow
